kotlin/samples/dungeon/common/src/main/java/com/squareup/sample/dungeon/GameWorkflow.kt

```diff
@@ -0,0 +1,244 @@
+import com.squareup.sample.dungeon.GameWorkflow.State
+import com.squareup.sample.dungeon.board.Board
+import com.squareup.sample.dungeon.board.Board.Location
+import com.squareup.workflow.RenderContext
+import com.squareup.workflow.Snapshot
+import com.squareup.workflow.StatefulWorkflow
+import com.squareup.workflow.Worker
+import com.squareup.workflow.WorkflowAction.Companion.enterState
+import com.squareup.workflow.onWorkerOutput
+import com.squareup.workflow.renderChild
+import kotlinx.coroutines.delay
+import kotlin.math.roundToLong
+import kotlin.random.Random
+
+private val PAUSED_TICKER = Worker.createSideEffect(key = "noop") {}
+
+class GameWorkflow(
+  private val playerWorkflow: PlayerWorkflow,
+  private val aiWorkflows: List<ActorWorkflow>,
+  private val random: Random
+) : StatefulWorkflow<Input, State, Output, GameRendering>() {
+
+  /**
+   * @param board Should not change while the game is running.
+   */
+  data class Input(
+    val board: Board,
+    val ticksPerSecond: Int = 15
+  )
+
+  /**
+   * @param finishedSnapshot If non-null, the game is finished and this was the last rendering
+   * before the game finished.
+   */
+  data class State(
+    val game: Game
+  )
+
+  sealed class Output {
+    /**
+     * Emitted by [GameWorkflow] if the controller should be vibrated.
+     */
+    object Vibrate : Output()
+
+    object PlayerWasEaten : Output()
+  }
+
+  data class GameRendering(
+    val board: Board,
+    val onPlayerEvent: ((PlayerWorkflow.Event) -> Unit)?
+  )
+
+  override fun initialState(
+    input: Input,
+    snapshot: Snapshot?
+  ): State {
+    val board = input.board
+    return State(game = Game(
+        playerLocation = random.nextEmptyLocation(board),
+        aiLocations = aiWorkflows.map { random.nextEmptyLocation(board) }
+    ))
+  }
+
+  override fun onInputChanged(
+    old: Input,
+    new: Input,
+    state: State
+  ): State {
+    check(old.board == new.board) { "Expected board to not change during the game." }
+    return state
+  }
+
+  override fun render(
+    input: Input,
+    state: State,
+    context: RenderContext<State, Output>
+  ): GameRendering {
+    val running = !state.game.isPlayerEaten
+    // Stop actors from ticking if the game is paused or finished.
+    val ticker = if (running) createTickerWorker(input.ticksPerSecond) else PAUSED_TICKER
+    val game = state.game
+    val board = input.board
+
+    // Render the player.
+    val playerInput = ActorInput(board, game.playerLocation, ticker)
+    val playerRendering = context.renderChild(playerWorkflow, playerInput)
+
+    // Render all the other actors.
+    val aiRenderings = aiWorkflows.zip(game.aiLocations)
+        .mapIndexed { index, (aiWorkflow, aiLocation) ->
+          val aiInput = ActorInput(board, aiLocation, ticker)
+          aiLocation to context.renderChild(aiWorkflow, aiInput, key = index.toString())
+        }
+
+    // If the game is paused or finished, just render the board without ticking.
```

Comment:
nit: this kind of thing is a little easier to read if the trivial case is first. That is, `if (!running)…` — makes your comment easier to follow too.